ruby/ql/src/queries/security/cwe-078/NonConstantKernelOpen.ql

```diff
@@ -18,23 +18,33 @@
 import codeql.ruby.security.KernelOpenQuery
 import codeql.ruby.AST
 import codeql.ruby.ApiGraphs
+import codeql.ruby.DataFlow
 
 from AmbiguousPathCall call
 where
-  not hasConstantPrefix(call.getPathArgument().getALocalSource().asExpr().getExpr()) and
+  call.getNumberOfArguments() > 0 and
+  not hasConstantPrefix(call.getPathArgument()) and
   not call.getPathArgument().getALocalSource() =
     API::getTopLevelMember("File").getAMethodCall("join")
 select call,
   "Call to " + call.getName() + " with a non-constant value. Consider replacing it with " +
     call.getReplacement() + "."
 
-predicate hasConstantPrefix(Expr e) {
+predicate hasConstantPrefix(DataFlow::Node node) {
+  hasConstantPrefix(node.getALocalSource())
+  or
   // if it's a format string, then the first argument is not a constant string
-  e.(StringlikeLiteral).getComponent(0) instanceof StringTextComponent
+  node.asExpr().getExpr().(StringlikeLiteral).getComponent(0) instanceof StringTextComponent
   or
   // it is not a constant string argument
-  exists(e.getConstantValue())
+  exists(node.asExpr().getExpr().getConstantValue())
   or
   // not a concatenation that starts with a constant string
-  hasConstantPrefix(e.(AddExpr).getLeftOperand())
+  exists(DataFlow::ExprNode prefix |
+    node.asExpr().getExpr().(AddExpr).getLeftOperand() = prefix.asExpr().getExpr() and
+    hasConstantPrefix(prefix)
+  )
+  or
+  // is a .freeze call on a constant string
+  node.asExpr().getExpr().(ConstantReadAccess).getValue().(MethodCall).getMethodName() = "freeze"
```

Comment:
```suggestion
  exists(DataFlow::CallNode call | node = call and call.getMethodName() = "freeze" |
    hasConstantPrefix(call.getReceiver())
  )
  or
  // is a constant read of a constant string
  exists(DataFlow::Node constant |
    constant.asExpr().getExpr() = node.asExpr().getExpr().(ConstantReadAccess).getValue() and
    hasConstantPrefix(constant)
  )
```

I figured out why my approach didn't work.  
1: I forgot to mention `node` in my case for `freeze` :doh: 
2: We need a recursive case for `ConstantReadAccess`, which enables us to find the definition of `CONSTANT_WITH_FREEZE` from where it's used.